keanu-project/src/main/java/io/improbable/keanu/network/BayesianNetwork.java

```diff
@@ -18,15 +15,34 @@
 public class BayesianNetwork {
 
     private final List<? extends Vertex> vertices;
+    private final HashMap<String, Vertex> labelSet;
+
+    private void addLabels(Set<? extends Vertex>vertices) {
+        for (Vertex v : vertices) {
+            String label = v.getLabel();
+            if (label != null && labelSet.containsKey(label)) {
+                throw new IllegalArgumentException("Vertex Label Repeated: " + label);
+            } else {
+                labelSet.put(label, v);
+            }
+        }
+    }
 
     public BayesianNetwork(Set<? extends Vertex> vertices) {
         this.vertices = ImmutableList.copyOf(vertices);
+        this.labelSet = new HashMap<>();
```

Comment:
If you want to be more functional you could change your `addLabels(...)` to not have side-effects and be a pure static function that returns the label to Vertex map. `this.labelSet = getLabels(vertices)`. I tend to favour that approach but in this case it's a matter of preferred style.